In [1]:
from pathlib import Path
import pandas as pd

PROJECT_ROOT = Path.cwd()

if PROJECT_ROOT.name == "notebooks":
    PROJECT_ROOT = PROJECT_ROOT.parent

RAW_DATA_DIR = PROJECT_ROOT / "data" / "raw"

print("Project root:", PROJECT_ROOT)
print("Raw data folder:", RAW_DATA_DIR)
print("Files found:")

for file in RAW_DATA_DIR.glob("*.csv"):
    print("-", file.name)

Project root: c:\Users\PC\Documents\project-foresight
Raw data folder: c:\Users\PC\Documents\project-foresight\data\raw
Files found:
- inventory_daily.csv
- products_master.csv
- sales_transactions.csv


In [2]:
sales = pd.read_csv(RAW_DATA_DIR / "sales_transactions.csv")
products = pd.read_csv(RAW_DATA_DIR / "products_master.csv")
inventory = pd.read_csv(RAW_DATA_DIR / "inventory_daily.csv")

print("Sales shape:", sales.shape)
print("Products shape:", products.shape)
print("Inventory shape:", inventory.shape)

Sales shape: (357645, 11)
Products shape: (150, 9)
Inventory shape: (109650, 11)


In [3]:
print("Sales columns:")
print(sales.columns.tolist())

print("\nProducts columns:")
print(products.columns.tolist())

print("\nInventory columns:")
print(inventory.columns.tolist())

Sales columns:
['Invoice', 'InvoiceDate', 'CustomerID', 'StockCode', 'Description', 'Category', 'Quantity', 'UnitPrice', 'Country', 'Channel', 'TotalPrice']

Products columns:
['StockCode', 'Description', 'Category', 'UnitCost', 'UnitPrice', 'LeadTimeDays', 'ShelfLifeDays', 'Supplier', 'MinOrderQty']

Inventory columns:
['Date', 'StockCode', 'OpeningStock', 'UnitsSold', 'LostSales', 'ClosingStock', 'OnOrder', 'ReorderPlaced', 'Stockout', 'ReorderPoint', 'TargetStock']


In [4]:
print("Sales sample:")
display(sales.head())

print("Products sample:")
display(products.head())

print("Inventory sample:")
display(inventory.head())

Sales sample:


,Invoice,InvoiceDate,CustomerID,StockCode,Description,Category,Quantity,UnitPrice,Country,Channel,TotalPrice
0,600000,2023-08-29 08:20:00,12001.0,SKU10000,WHITE HANGING HEART T-LIGHT HOLDER,Home Decor,97,10.30,United Kingdom,Web,999.10
1,600000,2023-08-29 08:20:00,12001.0,SKU10031,SPICE RACK WOODEN,Kitchenware,4,40.53,United Kingdom,Web,162.12
2,600000,2023-08-29 08:20:00,12001.0,SKU10141,JAM JAR STRAWBERRY,Food & Bev,1,7.01,United Kingdom,Web,7.01
3,600000,2023-08-29 08:20:00,12001.0,SKU10028,IVORY KNITTED MUG COSY,Kitchenware,1,27.75,United Kingdom,Web,27.75
4,600001,2024-03-28 17:08:00,12001.0,SKU10123,SHAMPOO HERBAL,Beauty,1,6.43,United Kingdom,Web,6.43


Products sample:


,StockCode,Description,Category,UnitCost,UnitPrice,LeadTimeDays,ShelfLifeDays,Supplier,MinOrderQty
0,SKU10000,WHITE HANGING HEART T-LIGHT HOLDER,Home Decor,6.59,10.30,7,365,SUP107,10
1,SKU10001,WHITE METAL LANTERN,Home Decor,8.37,15.53,7,365,SUP110,100
2,SKU10002,CREAM CUPID HEARTS COAT HANGER,Home Decor,4.47,9.34,7,365,SUP121,10
3,SKU10003,GLASS STAR FROSTED T-LIGHT HOLDER,Home Decor,14.97,24.29,6,365,SUP127,10
4,SKU10004,WALL MIRROR ROUND,Home Decor,9.16,16.01,9,365,SUP111,10


Inventory sample:


,Date,StockCode,OpeningStock,UnitsSold,LostSales,ClosingStock,OnOrder,ReorderPlaced,Stockout,ReorderPoint,TargetStock
0,2023-01-01,SKU10000,431,0,0,431,0,0,0,215,431
1,2023-01-02,SKU10000,431,0,0,431,0,0,0,215,431
2,2023-01-03,SKU10000,431,0,0,431,0,0,0,215,431
3,2023-01-04,SKU10000,431,0,0,431,0,0,0,215,431
4,2023-01-05,SKU10000,431,0,0,431,0,0,0,215,431


In [5]:
datasets = {
    "Sales": sales,
    "Products": products,
    "Inventory": inventory
}

for name, df in datasets.items():
    print("\n" + "=" * 60)
    print(name.upper())
    print("=" * 60)

    print("Shape:", df.shape)
    print("Duplicate rows:", df.duplicated().sum())

    print("\nData types:")
    print(df.dtypes)

    missing = df.isna().sum()
    missing = missing[missing > 0]

    print("\nMissing values:")
    if missing.empty:
        print("No missing values")
    else:
        display(
            pd.DataFrame({
                "Missing Count": missing,
                "Missing Percentage": (missing / len(df) * 100).round(2)
            })
        )


SALES
Shape: (357645, 11)
Duplicate rows: 0

Data types:
Invoice         object
InvoiceDate     object
CustomerID     float64
StockCode       object
Description     object
Category        object
Quantity         int64
UnitPrice      float64
Country         object
Channel         object
TotalPrice     float64
dtype: object

Missing values:


,Missing Count,Missing Percentage
CustomerID,17882,5.0



PRODUCTS
Shape: (150, 9)
Duplicate rows: 0

Data types:
StockCode         object
Description       object
Category          object
UnitCost         float64
UnitPrice        float64
LeadTimeDays       int64
ShelfLifeDays      int64
Supplier          object
MinOrderQty        int64
dtype: object

Missing values:
No missing values

INVENTORY
Shape: (109650, 11)
Duplicate rows: 0

Data types:
Date             object
StockCode        object
OpeningStock      int64
UnitsSold         int64
LostSales         int64
ClosingStock      int64
OnOrder           int64
ReorderPlaced     int64
Stockout          int64
ReorderPoint      int64
TargetStock       int64
dtype: object

Missing values:
No missing values


In [6]:
sales_dates = pd.to_datetime(sales["InvoiceDate"], errors="coerce")
invoice_text = sales["Invoice"].astype(str)

print("Sales date range:", sales_dates.min(), "to", sales_dates.max())
print("Invalid sales dates:", sales_dates.isna().sum())
print("Unique invoices:", sales["Invoice"].nunique())
print("Unique SKUs:", sales["StockCode"].nunique())
print("Unique customers:", sales["CustomerID"].nunique(dropna=True))

print("\nSales quality checks:")
print("Missing CustomerID:", sales["CustomerID"].isna().sum())
print("Missing StockCode:", sales["StockCode"].isna().sum())
print("Quantity <= 0:", (sales["Quantity"] <= 0).sum())
print("UnitPrice <= 0:", (sales["UnitPrice"] <= 0).sum())
print(
    "Invoice numbers beginning with C:",
    invoice_text.str.startswith("C", na=False).sum()
)
print("Exact duplicate sales rows:", sales.duplicated().sum())

Sales date range: 2023-01-10 15:42:00 to 2024-12-30 20:00:00
Invalid sales dates: 0
Unique invoices: 91127
Unique SKUs: 150
Unique customers: 7603

Sales quality checks:
Missing CustomerID: 17882
Missing StockCode: 0
Quantity <= 0: 10750
UnitPrice <= 0: 0
Invoice numbers beginning with C: 10750
Exact duplicate sales rows: 0


In [7]:
sales_skus = set(sales["StockCode"].dropna().astype(str))
product_skus = set(products["StockCode"].dropna().astype(str))

missing_product_skus = sales_skus - product_skus

print("Product records:", len(products))
print("Unique product SKUs:", products["StockCode"].nunique())
print("Duplicate StockCode values:", products["StockCode"].duplicated().sum())
print("Missing descriptions:", products["Description"].isna().sum())
print("Missing categories:", products["Category"].isna().sum())
print("UnitCost <= 0:", (products["UnitCost"] <= 0).sum())
print("UnitPrice <= 0:", (products["UnitPrice"] <= 0).sum())
print("LeadTimeDays <= 0:", (products["LeadTimeDays"] <= 0).sum())
print("Sales SKUs missing from product master:", len(missing_product_skus))

if missing_product_skus:
    print("Missing SKU examples:", list(missing_product_skus)[:10])

Product records: 150
Unique product SKUs: 150
Duplicate StockCode values: 0
Missing descriptions: 0
Missing categories: 0
UnitCost <= 0: 0
UnitPrice <= 0: 0
LeadTimeDays <= 0: 0
Sales SKUs missing from product master: 0


## Day 2 Data-Profiling Findings

### Sales Transactions

- The sales dataset contains 357,645 rows and 11 columns.
- It covers the period from 10 January 2023 to 30 December 2024.
- It contains 91,127 unique invoices, 150 SKUs and 7,603 identified customers.
- CustomerID is missing in 17,882 records, representing 5% of the dataset.
- There are 10,750 records with non-positive quantities.
- The same 10,750 records belong to invoices beginning with C, indicating cancellations.
- No invalid or non-positive unit prices were found.
- No exact duplicate sales rows were found.

### Product Master

- The product master contains 150 unique products.
- No duplicate product IDs were found.
- No missing descriptions or categories were found.
- No invalid unit costs, selling prices or lead times were found.
- Every sales SKU exists in the product master.

### Inventory

- The inventory dataset contains 109,650 rows and covers 1 January 2023 to 31 December 2024.
- It contains 150 unique SKUs.
- No invalid dates or duplicate Date-SKU records were found.
- Every inventory SKU exists in the product master.
- No negative numeric values or missing values were found.

### Proposed Cleaning Decisions

- Exclude cancellation transactions from fulfilled-demand calculations.
- Retain rows with missing CustomerID because customer identity is not required for SKU-level demand forecasting.
- Convert date columns to datetime format through code.
- Standardise StockCode values as strings.
- Preserve the original raw files without manual editing.

In [8]:
inventory_dates = pd.to_datetime(inventory["Date"], errors="coerce")

product_skus = set(products["StockCode"].dropna().astype(str))
inventory_skus = set(inventory["StockCode"].dropna().astype(str))

missing_inventory_skus = inventory_skus - product_skus

print("Inventory date range:", inventory_dates.min(), "to", inventory_dates.max())
print("Invalid inventory dates:", inventory_dates.isna().sum())
print("Unique inventory SKUs:", inventory["StockCode"].nunique())

print(
    "Duplicate Date-SKU records:",
    inventory.duplicated(subset=["Date", "StockCode"]).sum()
)

print(
    "Inventory SKUs missing from product master:",
    len(missing_inventory_skus)
)

numeric_columns = inventory.select_dtypes(include="number").columns
negative_counts = (inventory[numeric_columns] < 0).sum()
negative_counts = negative_counts[negative_counts > 0]

print("\nNegative-value checks:")
if negative_counts.empty:
    print("No negative numeric values")
else:
    display(negative_counts.to_frame("Negative Count"))

Inventory date range: 2023-01-01 00:00:00 to 2024-12-31 00:00:00
Invalid inventory dates: 0
Unique inventory SKUs: 150
Duplicate Date-SKU records: 0
Inventory SKUs missing from product master: 0

Negative-value checks:
No negative numeric values
